pkg/apis/alertmanager/v1alpha1/alertmanagerconfiguration_types.go

```diff
@@ -0,0 +1,551 @@
+	}
+	if len(headers) == 0 {
+		headers = nil
+	}
+
+	return alertmanagerconfig.EmailConfig{
+		NotifierConfig: alertmanagerconfig.NotifierConfig{VSendResolved: e.SendResolved},
+		To:             e.To,
+		From:           e.From,
+		Hello:          e.Hello,
+		Smarthost:      e.Smarthost,
+		AuthUsername:   e.AuthUsername,
+		AuthPassword:   authPassword,
+		AuthSecret:     authSecret,
+		AuthIdentity:   e.AuthIdentity,
+		Headers:        headers,
+		HTML:           e.HTML,
+		Text:           e.Text,
+		RequireTLS:     e.RequireTLS,
+	}
+}
+
+// PagerDutyReceiver holds the configuration for a PagerDuty receiver
+type PagerDutyReceiver struct {
+	// Whether to send resolved notifications or not.
+	SendResolved bool `json:"sendResolved,omitempty"`
+
+	// PagerDuty integration key (when using Events API
+	// v2). Either this key or service_key needs to be defined.
+	RoutingKey *corev1.SecretKeySelector `json:"routingKey,omitempty"`
+
+	// The URL to send requests to.
+	// +kubebuilder:validation:Pattern=`^(http|https)://.+$`
+	URL string `json:"url,omitempty"`
+
+	// Client identification
+	Client string `json:"client,omitempty"`
+
+	// Backlink to the sendor of notification.
+	// +kubebuilder:validation:Pattern=`^(http|https)://.+$`
+	ClientURL string `json:"clientUrl,omitempty"`
+
+	// Description of the incident.
+	Description string `json:"description,omitempty"`
+
+	// Severity of the incident.
+	Severity string `json:"severity,omitempty"`
+
+	// Arbitrary key/value pairs that provide further detail about
+	// the incident.
+	Details []KeyValue `json:"details,omitempty"`
+
+	// TODO: enable this
+	// HTTP client configuration.
+	// HTTPConfig *HTTPConfig `json:"httpConfig,omitempty"`
+}
+
+// toAMPagerdutyConfig converts a PagerDutyReceiver specified in a Receiver in
+// an AlertManagerConfiguration CR, to the type expected for serialization into
+// the Alertmanager YAML config file.
+func (p PagerDutyReceiver) toAMPagerdutyConfig(
+	amcObjectMeta metav1.ObjectMeta,
+	getValueFromSecretKeySelector func(namespace string, secretKeySelector *corev1.SecretKeySelector) (string, error),
+) alertmanagerconfig.PagerdutyConfig {
+
+	routingKey := ""
+	if p.RoutingKey != nil {
+		secretRoutingKey, err := getValueFromSecretKeySelector(amcObjectMeta.Namespace, p.RoutingKey)
+		if err == nil {
+			routingKey = secretRoutingKey
+		}
+	}
+
+	details := map[string]string{}
+	for _, detail := range p.Details {
+		details[detail.Key] = detail.Value
+	}
+	if len(details) == 0 {
+		details = nil
+	}
+
+	return alertmanagerconfig.PagerdutyConfig{
+		NotifierConfig: alertmanagerconfig.NotifierConfig{VSendResolved: p.SendResolved},
+		RoutingKey:     routingKey,
+		URL:            p.URL,
+		Client:         p.Client,
+		ClientURL:      p.ClientURL,
+		Description:    p.Description,
+		Details:        details,
+		Severity:       p.Severity,
+
+		// TODO: are these properties used?
+		Class:     "",
+		Component: "",
+		Group:     "",
```

Comment:
The reason this is a TODO is that these fields weren't in the prometheus-operator proposal, and they weren't documented in the Alertmanager docs. I've created a PR to those docs to get them included.

I'll update this to make them available in the CRD, so there's one less TODO :slightly_smiling_face: